### ISSUE TITLE
denonavr not working after firmware upgrade on Denon AVC X3700H

### ISSUE THREAD

**Issue by flokru** [ol-iver/denonavr#202]
I recently installed a new firmware on my Denon AVC X3700H. After this update, denonavr is no longer working.

Model: Denon AVC X3700H
Firmware Version: 6220-9043-7161-3085 (DTS Version | 3.90.50.54)

Python version: 3.8.3

This is what happens with denonavr:

```
In [1]: import denonavr

In [2]: d = denonavr.DenonAVR("192.168.0.12", "Denon")
   ...: d.update()
   ...:
   ...:

---------------------------------------------------------------------------
RemoteProtocolError                       Traceback (most recent call last)
~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/httpx/_exceptions.py in map_exceptions(mapping, **kwargs)
    325     try:
--> 326         yield
    327     except Exception as exc:

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/httpx/_client.py in _send_single_request(self, request, timeout)
   1491         with map_exceptions(HTTPCORE_EXC_MAP, request=request):
-> 1492             (status_code, headers, stream, ext) = await transport.arequest(
   1493                 request.method.encode(),

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/httpx/_transports/default.py in arequest(self, method, url, headers, stream, ext)
    168     ) -> typing.Tuple[int, Headers, httpcore.AsyncByteStream, dict]:
--> 169         return await self._pool.arequest(
    170             method, url, headers=headers, stream=stream, ext=ext

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/httpcore/_async/connection_pool.py in arequest(self, method, url, headers, stream, ext)
    217             try:
--> 218                 response = await connection.arequest(
    219                     method, url, headers=headers, stream=stream, ext=ext

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/httpcore/_async/connection.py in arequest(self, method, url, headers, stream, ext)
    105         )
--> 106         return await self.connection.arequest(method, url, headers, stream, ext)
    107

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/httpcore/_async/http11.py in arequest(self, method, url, headers, stream, ext)
     71             headers,
---> 72         ) = await self._receive_response(timeout)
     73         response_stream = AsyncIteratorByteStream(

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/httpcore/_async/http11.py in _receive_response(self, timeout)
    132         while True:
--> 133             event = await self._receive_event(timeout)
    134             if isinstance(event, h11.Response):

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/httpcore/_async/http11.py in _receive_event(self, timeout)
    168             with map_exceptions({h11.RemoteProtocolError: RemoteProtocolError}):
--> 169                 event = self.h11_state.next_event()
    170

~/.pyenv/versions/3.8.3/lib/python3.8/contextlib.py in __exit__(self, type, value, traceback)
    130             try:
--> 131                 self.gen.throw(type, value, traceback)
    132             except StopIteration as exc:

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/httpcore/_exceptions.py in map_exceptions(map)
     11             if isinstance(exc, from_exc):
---> 12                 raise to_exc(exc) from None
     13         raise

RemoteProtocolError: can't handle event type ConnectionClosed when role=SERVER and state=SEND_RESPONSE

The above exception was the direct cause of the following exception:

RemoteProtocolError                       Traceback (most recent call last)
<ipython-input-2-0a32a765050b> in <module>
      1 d = denonavr.DenonAVR("192.168.0.12", "Denon")
----> 2 d.update()
      3

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/denonavr/decorators.py in wrapper(*args, **kwargs)
    138
    139             try:
--> 140                 return loop.run_until_complete(async_func(*args, **kwargs))
    141             finally:
    142                 loop.close()

~/.pyenv/versions/3.8.3/lib/python3.8/asyncio/base_events.py in run_until_complete(self, future)
    614             raise RuntimeError('Event loop stopped before Future completed.')
    615
--> 616         return future.result()
    617
    618     def stop(self):

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/denonavr/denonavr.py in async_update(self)
    155         # Ensure that the device is setup
    156         if self._is_setup is False:
--> 157             await self.async_setup()
    158
    159         # Create a cache id for this global update

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/denonavr/denonavr.py in async_setup(self)
    130         """Ensure that configuration is loaded from receiver asynchronously."""
    131         # Device setup
--> 132         await self._device.async_setup()
    133         if self._name is None:
    134             self._name = self._device.friendly_name

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/denonavr/foundation.py in async_setup(self)
     95         """Ensure that configuration is loaded from receiver asynchronously."""
     96         # Own setup
---> 97         await self.async_identify_receiver()
     98         await self.async_identify_update_method()
     99         await self.async_get_device_info()

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/denonavr/foundation.py in async_identify_receiver(self)
    129             try:
    130                 # Deviceinfo.xml is static and can be cached for the whole time
--> 131                 xml = await self.api.async_get_xml(
    132                     self.urls.deviceinfo, cache_id=id(self))
    133             except (AvrTimoutError, AvrNetworkError) as err:

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/denonavr/decorators.py in wrapper(*args, **kwargs)
     88     async def wrapper(*args, **kwargs):
     89         try:
---> 90             return await func(*args, **kwargs)
     91         except Exception as err:
     92             _LOGGER.debug("Exception %s raised, clearing cache", err)

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/async_lru.py in wrapped(*fn_args, **fn_kwargs)
    235
    236             _cache_miss(wrapped, key)
--> 237             return (yield from asyncio.shield(fut, loop=_loop))
    238
    239         _cache_clear(wrapped)

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/denonavr/decorators.py in wrapper(*args, **kwargs)
     42     async def wrapper(*args, **kwargs):
     43         try:
---> 44             return await func(*args, **kwargs)
     45         except httpx.HTTPStatusError as err:
     46             _LOGGER.debug(

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/denonavr/api.py in async_get_xml(self, request, cache_id)
    137         """Return XML data from HTTP GET endpoint asynchronously."""
    138         # HTTP GET to endpoint
--> 139         res = await self.async_get(request)
    140         # create ElementTree
    141         xml_root = fromstring(res.text)

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/denonavr/decorators.py in wrapper(*args, **kwargs)
     42     async def wrapper(*args, **kwargs):
     43         try:
---> 44             return await func(*args, **kwargs)
     45         except httpx.HTTPStatusError as err:
     46             _LOGGER.debug(

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/denonavr/api.py in async_get(self, request, port)
     83         client = self.async_client_getter()
     84         try:
---> 85             res = await client.get(endpoint, timeout=self.timeout)
     86             res.raise_for_status()
     87         finally:

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/httpx/_client.py in get(self, url, params, headers, cookies, auth, allow_redirects, timeout)
   1537         **Parameters**: See `httpx.request`.
   1538         """
-> 1539         return await self.request(
   1540             "GET",
   1541             url,

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/httpx/_client.py in request(self, method, url, content, data, files, json, params, headers, cookies, auth, allow_redirects, timeout)
   1359             cookies=cookies,
   1360         )
-> 1361         response = await self.send(
   1362             request, auth=auth, allow_redirects=allow_redirects, timeout=timeout
   1363         )

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/httpx/_client.py in send(self, request, stream, auth, allow_redirects, timeout)
   1394         auth = self._build_request_auth(request, auth)
   1395
-> 1396         response = await self._send_handling_auth(
   1397             request,
   1398             auth=auth,

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/httpx/_client.py in _send_handling_auth(self, request, auth, timeout, allow_redirects, history)
   1432
   1433         while True:
-> 1434             response = await self._send_handling_redirects(
   1435                 request,
   1436                 timeout=timeout,

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/httpx/_client.py in _send_handling_redirects(self, request, timeout, allow_redirects, history)
   1464                 )
   1465
-> 1466             response = await self._send_single_request(request, timeout)
   1467             response.history = list(history)
   1468

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/httpx/_client.py in _send_single_request(self, request, timeout)
   1490
   1491         with map_exceptions(HTTPCORE_EXC_MAP, request=request):
-> 1492             (status_code, headers, stream, ext) = await transport.arequest(
   1493                 request.method.encode(),
   1494                 request.url.raw,

~/.pyenv/versions/3.8.3/lib/python3.8/contextlib.py in __exit__(self, type, value, traceback)
    129                 value = type()
    130             try:
--> 131                 self.gen.throw(type, value, traceback)
    132             except StopIteration as exc:
    133                 # Suppress StopIteration *unless* it's the same exception that

~/.pyenv/versions/3.8.3/envs/p8-base/lib/python3.8/site-packages/httpx/_exceptions.py in map_exceptions(mapping, **kwargs)
    341
    342         message = str(exc)
--> 343         raise mapped_exc(message, **kwargs) from exc  # type: ignore
    344
    345

RemoteProtocolError: can't handle event type ConnectionClosed when role=SERVER and state=SEND_RESPONSE
```

**Comment by ol-iver** [comment]
Did you try a power cycle of your receiver? Some receivers seem to deactivate the API for unknown reason after a while and unplugging it could help.
If this does not help, it might be related to the firmware update. Could you please run [this script](https://github.com/scarface-4711/misc_python_tools/tree/master/denonavr) and upload the XML files it generates here? They include the responses of the API calls of this library.

**Comment by flokru** [comment]
Power cycling did not help. I ran the script:

```
$ python3 denon_receiver_xml.py --host 192.168.0.12 --prefix AVC-X3700H
HTTP Status Code of /goform/AppCommand.xml: 403
HTTP Status Code of /goform/AppCommand.xml: 403
HTTP Status Code of /goform/Deviceinfo.xml: 403
HTTP Status Code of /goform/formMainZone_MainZoneXmlStatus.xml: 403
HTTP Status Code of /goform/formMainZone_MainZoneXml.xml: 403
HTTP Status Code of /goform/formNetAudio_StatusXml.xml: 403
HTTP Status Code of /goform/formTuner_TunerXml.xml: 403
HTTP Status Code of /goform/formTuner_HdXml.xml: 403
HTTP Status Code of /goform/formZone2_Zone2XmlStatus.xml: 403
HTTP Status Code of /goform/formZone3_Zone3XmlStatus.xml: 403
HTTP Status Code of /goform/AppCommand.xml: 200
HTTP Status Code of /goform/AppCommand.xml: 200
HTTP Status Code of /goform/Deviceinfo.xml: 200
HTTP Status Code of /goform/formMainZone_MainZoneXmlStatus.xml: 403
HTTP Status Code of /goform/formMainZone_MainZoneXml.xml: 403
HTTP Status Code of /goform/formNetAudio_StatusXml.xml: 403
HTTP Status Code of /goform/formTuner_TunerXml.xml: 403
HTTP Status Code of /goform/formTuner_HdXml.xml: 403
HTTP Status Code of /goform/formZone2_Zone2XmlStatus.xml: 403
HTTP Status Code of /goform/formZone3_Zone3XmlStatus.xml: 403
```

Files are here:

[Archive 2.zip](https://github.com/scarface-4711/denonavr/files/6413081/Archive.2.zip)

For the record, the receiver was in standby while running the command (I am currently not near the device to turn it on…).

**Comment by ol-iver** [comment]
Thanks for the files. Your issue should be fixed now in version 0.10.7.

**Comment by flokru** [comment]
It works! Thanks for the fix!
